torch/distributed/_functional_collectives.py

```diff
@@ -546,6 +553,582 @@ def all_to_all_single_autograd(
+    ctx.group_name = group_name
+    ctx.group_size = group_size
+
+
+torch.library.register_autograd(
+    "_c10d_functional::all_gather_into_tensor_coalesced",
+    all_gather_into_tensor_coalesced_backward,
+    setup_context=all_gather_into_tensor_coalesced_setup_context,
+)
+
+
+def reduce_scatter_tensor_coalesced_backward(ctx, grad_outputs: list[torch.Tensor]):
+    """
+    Backward for reduce_scatter_tensor_coalesced: all_gather each gradient.
+
+    Forward reduces and scatters tensors to ranks, backward gathers gradients
+    from all ranks.
+
+    Args:
+        ctx: Context object with group_name, group_size, and reduce_op
+        grad_outputs: Gradients from downstream operations (one per input tensor)
+
+    Returns:
+        Tuple of (grad_inputs..., grad_reduce_op, grad_group_size, grad_group_name)
+        grad_reduce_op, grad_group_size, grad_group_name are None (not differentiable)
+    """
+    group_name = ctx.group_name
+    group_size = ctx.group_size
+    reduce_op = ctx.reduce_op
+
+    # Lazy validation: check reduce_op only when backward is called
+    if reduce_op != "sum":
+        raise RuntimeError(
+            f"reduce_scatter_tensor_coalesced backward only supports 'sum' reduction, got '{reduce_op}'"
+        )
+
+    # Backward does all_gather on list of gradients
+    grad_inputs = torch.ops._c10d_functional.all_gather_into_tensor_coalesced(
+        [grad_output.contiguous() for grad_output in grad_outputs],
+        group_size,
+        group_name,
+    )
+    return (list(map(wait_tensor, grad_inputs)), None, None, None)
+
+
+def reduce_scatter_tensor_coalesced_setup_context(ctx, inputs, output):
+    """
+    Setup context for reduce_scatter_tensor_coalesced backward.
+
+    Args:
+        ctx: Context object to save state for backward
+        inputs: Tuple of (tensor_list, reduce_op, group_size, group_name)
+        output: Output from forward pass
+    """
+    tensor_list, reduce_op, group_size, group_name = inputs
+    ctx.group_name = group_name
+    ctx.group_size = group_size
+    ctx.reduce_op = reduce_op.lower()
+
+
+torch.library.register_autograd(
+    "_c10d_functional::reduce_scatter_tensor_coalesced",
+    reduce_scatter_tensor_coalesced_backward,
+    setup_context=reduce_scatter_tensor_coalesced_setup_context,
+)
+
+
+@torch.library.custom_op("_c10d_functional::all_reduce_sum_invariant", mutates_args=())
+def all_reduce_sum_invariant_op(input: torch.Tensor, group_name: str) -> torch.Tensor:
+    """
+    This is the low-level implementation. Users should call the wrapper function.
+    Args:
+        input: Input tensor
+        group_name: Process group name string
+
+    Returns:
+        Reduced tensor
+    """
+    output = torch.ops._c10d_functional.all_reduce(input, "sum", group_name)
+    if not input.requires_grad:
+        return _maybe_wrap_tensor(output)
+
+    return wait_tensor(output)
+
+
+@all_reduce_sum_invariant_op.register_fake
```

Comment:
nit: the code organization isn't consistent.  For all_reduce_sum_invariant, the op def, fake, backward, and setup_ctx are all together.  For other ops, the op def + fake are above, and the backward+setup ctx are below.

To @ezyang's point i'd avoid reordering existing functions for now in this PR, if you want to do that then do it in a separate PR.  But for now could we follow the existing structure?